meta/autoid/autoid.go

```diff
@@ -219,132 +221,180 @@ func (alloc *allocator) Rebase(tableID, requiredBase int64, allocIDs bool) error
+	}
+	if N == 0 {
+		return []int64{}, nil
+	}
+	alloc.mu.Lock()
+	defer alloc.mu.Unlock()
+	if alloc.isUnsigned {
+		return alloc.allocN4Unsigned(tableID, N)
+	}
+	return alloc.allocN4Signed(tableID, N)
+}
+
+func (alloc *allocator) allocN4Signed(tableID int64, N uint64) ([]int64, error) {
+	N1 := int64(N)
+	// Condition alloc.base+N1 > alloc.end will overflow when alloc.base + N1 > MaxInt64. So need this.
+	if math.MaxInt64-alloc.base < N1 {
+		return nil, ErrAutoincReadFailed
+	}
+	// The local rest is not enough for allocN, skip it.
+	if alloc.base+N1 > alloc.end {
 		var newBase, newEnd int64
 		startTime := time.Now()
+		// Although it may skip a segment here, we still think it is consumed.
 		consumeDur := startTime.Sub(alloc.lastAllocTime)
-		alloc.step = NextStep(alloc.step, consumeDur)
+		nextStep := NextStep(alloc.step, consumeDur)
+		// Make sure nextStep is big enough.
+		if nextStep <= N1 {
+			alloc.step = mathutil.MinInt64(N1*2, maxStep)
+		} else {
+			alloc.step = nextStep
+		}
 		err := kv.RunInNewTxn(alloc.store, true, func(txn kv.Transaction) error {
 			m := meta.NewMeta(txn)
 			var err1 error
 			newBase, err1 = m.GetAutoTableID(alloc.dbID, tableID)
 			if err1 != nil {
 				return err1
 			}
-			tmpStep := int64(mathutil.MinUint64(math.MaxUint64-uint64(newBase), uint64(alloc.step)))
+			tmpStep := mathutil.MinInt64(math.MaxInt64-newBase, alloc.step)
+			// The global rest is not enough for allocN.
+			if tmpStep < N1 {
+				return ErrAutoincReadFailed
+			}
 			newEnd, err1 = m.GenAutoTableID(alloc.dbID, tableID, tmpStep)
 			return err1
 		})
 		metrics.AutoIDHistogram.WithLabelValues(metrics.TableAutoIDAlloc, metrics.RetLabel(err)).Observe(time.Since(startTime).Seconds())
 		if err != nil {
-			return 0, err
+			return nil, err
 		}
 		alloc.lastAllocTime = time.Now()
-		if uint64(newBase) == math.MaxUint64 {
-			return 0, ErrAutoincReadFailed
+		if newBase == math.MaxInt64 {
+			return nil, ErrAutoincReadFailed
 		}
 		alloc.base, alloc.end = newBase, newEnd
 	}
-
-	if uint64(alloc.base)+uint64(1) == math.MaxUint64 {
-		return 0, ErrAutoincReadFailed
-	}
-	alloc.base = int64(uint64(alloc.base) + 1)
-	logutil.BgLogger().Debug("alloc unsigned ID",
-		zap.Uint64("ID", uint64(alloc.base)),
+	logutil.Logger(context.TODO()).Debug("alloc N signed ID",
+		zap.Uint64("from ID", uint64(alloc.base)),
+		zap.Uint64("to ID", uint64(alloc.base+N1)),
 		zap.Int64("table ID", tableID),
 		zap.Int64("database ID", alloc.dbID))
-	return alloc.base, nil
+	resN := make([]int64, 0, N1)
+	for i := alloc.base + 1; i <= alloc.base+N1; i++ {
+		// fix bug : maxInt64 will be allocated
+		if i == math.MaxInt64 {
+			return nil, ErrAutoincReadFailed
+		}
+		resN = append(resN, i)
+	}
+	alloc.base += N1
+	return resN, nil
 }
 
-func (alloc *allocator) alloc4Signed(tableID int64) (int64, error) {
-	if alloc.base == alloc.end { // step
+func (alloc *allocator) allocN4Unsigned(tableID int64, N uint64) ([]int64, error) {
+	N1 := int64(N)
```

Comment:
```suggestion
	n1 := int64(n)
```
Please rename all 'N' related variables to lowercase